bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java

```diff
@@ -0,0 +1,595 @@
+     * updates
+     */
+    private void msgLoop() {
+        try {
+            JsonObject hubMsg;
+            JsonPrimitive jsonService;
+            String service;
+
+            // Main message loop listening for updates from the hub
+            logger.debug("Starting message loop");
+            while (!shouldStop()) {
+                if (!connect()) {
+                    int sleep = hubReconnectSleep;
+                    logger.debug("Waiting {} seconds before re-attempting to connect.", sleep);
+                    if (hubReconnectSleep < HUB_RECONNECT_SLEEP_MAXIMUM) {
+                        hubReconnectSleep = hubReconnectSleep * 2; // Increase sleep time exponentially
+                    }
+                    restartMsgLoop(sleep);
+                    return;
+                } else {
+                    hubReconnectSleep = HUB_RECONNECT_SLEEP_MINIMUM; // Reset
+                }
+
+                if ((hubMsg = getHubMsg()) == null) {
+                    continue;
+                }
+
+                // Process message based on service type
+                if ((jsonService = hubMsg.getAsJsonPrimitive(HUB_TOKEN_SERVICE)) != null) {
+                    service = jsonService.getAsString();
+                } else {
+                    continue; // Ignore messages that don't have a service specified
+                }
+
+                if (service.equals(HUB_SERVICE_REPORT_ZONE_PROPERTIES)) {
+                    processMsgReportZoneProperties(hubMsg);
+                } else if (service.equals(HUB_SERVICE_ZONE_PROPERTIES_CHANGED)) {
+                    processMsgZonePropertiesChanged(hubMsg);
+                } else if (service.equals(HUB_SERVICE_LIST_ZONE)) {
+                    processMsgListZone(hubMsg);
+                } else if (service.equals(HUB_SERVICE_SYSTEM_INFO)) {
+                    processMsgSystemInfo(hubMsg);
+                }
+            }
+        } catch (RuntimeException e) {
+            logger.error("Hub controller failed", e);
+        }
+
+        // Shut down
+        disconnect();
+
+        // If there are still pending commands we need to cancel them
+        synchronized (stateCommandsLock)
+
+        {
+            stateCommands.forEach((zoneId, stateCommand) -> stateCommand.cancel(false));
+            stateCommands.clear();
+        }
+        synchronized (zoneCommandLock) {
+            CompletableFuture<List<Integer>> zoneCommand = this.zoneCommand;
+            if (zoneCommand != null) {
+                zoneCommand.cancel(false);
+                this.zoneCommand = null;
+            }
+        }
+        synchronized (macAddressCommandLock) {
+            CompletableFuture<String> macAddressCommand = this.macAddressCommand;
+            if (macAddressCommand != null) {
+                macAddressCommand.cancel(false);
+                this.macAddressCommand = null;
+            }
+        }
```

Comment:
As an example of using an `AtomicReference<@Nullable CompletableFuture<String>>, this code could be changed to:
```suggestion
            CompletableFuture<String> macAddressCommand = this.macAddressCommandRef.getAndSet(null);
            if (macAddressCommand != null) {
                macAddressCommand.cancel(false);
            }
```
If you added this utility method:
```
private void cancel(boolean interrupt, @Nullable Future<?> future){
    if(future != null){
         future.cancel(interrupt);
    }
}
```
Then the suggestion above becomes even easier.
```suggestion
            cancel(false, this.macAddressCommandRef.getAndSet(null));
```

The cancel utility method can be used for cancelling other futures as well and saves a tiny bit of code. If used enough the refactoring becomes worth it.

Reply:
See my notes above why I didn't remove the macAddressCommandLock. For this section an AtomicReference works well, but I found it not to work well for the others.